Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/RiskLevelChangeDetector.kt

```diff
@@ -65,9 +67,53 @@ class RiskLevelChangeDetector @Inject constructor(
 
         val oldRiskState = oldResult.riskState
         val newRiskState = newResult.riskState
-
         Timber.d("Last state was $oldRiskState and current state is $newRiskState")
 
+        // Check sending a notification when risk level changes
+        checkSendingNotification(oldRiskState, newRiskState)
+
+        // Save Survey related data based on the risk state
+        saveSurveyRiskState(oldRiskState, newRiskState, newResult)
+
+        // Save TestDonor risk level timestamps
+        saveTestDonorRiskLevelAnalytics(newResult)
+    }
+
+    private fun saveTestDonorRiskLevelAnalytics(
+        newRiskState: RiskLevelResult
+    ) {
+        // Save riskLevelTurnedRedTime if not already set before for high risk detection
+        Timber.i("riskLevelTurnedRedTime:%s", testResultDonorSettings.riskLevelTurnedRedTime.value)
+        if (testResultDonorSettings.riskLevelTurnedRedTime.value == null) {
+            if (newRiskState.isIncreasedRisk) {
+                testResultDonorSettings.riskLevelTurnedRedTime.update {
+                    newRiskState.calculatedAt
+                }
+                Timber.i(
+                    "newRiskState:%s, riskLevelTurnedRedTime:%s",
+                    newRiskState.riskState,
+                    newRiskState.calculatedAt
+                )
+            }
+        }
+
+        // Save most recent date of high or low risks
+        if (newRiskState.riskState in listOf(RiskState.INCREASED_RISK, RiskState.LOW_RISK)) {
+            Timber.i(
+                "newRiskState:%s, mostRecentDateWithHighOrLowRiskLevel:%s",
+                newRiskState.riskState,
+                newRiskState.calculatedAt
+            )
+            testResultDonorSettings.mostRecentDateWithHighOrLowRiskLevel.update {
+                newRiskState.calculatedAt
```

Comment:
This is the time when this result was calculated, which happens 5-6 times per day, but I think the value we want here is the day (e.g. 3 days ago) when we had an encounter.

For `riskLevelTurnedRedTime` (above) this works because it's only read and written once, the first time it happens.

But for `mostRecentDateWithHighOrLowRiskLevel` it will happen every day (hopefully), as `newRiskState.riskState` will be `FAILED/LOW/HIGH`.

We probably want this from `RiskLevelResult`:

```kotlin
    val lastRiskEncounterAt: Instant?
        get() = if (isIncreasedRisk) {
            aggregatedRiskResult?.mostRecentDateWithHighRisk
        } else {
            aggregatedRiskResult?.mostRecentDateWithLowRisk
        }
```

Reply:
which date do we rely on for  `the Home Screen Card ` ? I think this should be the one we use

Reply:
The homecards have multiple dates for calculation and encounters. I don't think the time this was calculated is interesting to analytics though, but rather (similar to the first value) it would be interesting for analytics when the user last had contact with someone that tested positive.

Reply:
I didn't notice this, but I agree it makes sense to use the first value when the increased risk was calculated for analytics.

Reply:
👍🏼